### PR TITLE
Reintroduce reverted commit "[clang-offload-bundler] Standardize TargetID field for bundler

### PR DESCRIPTION
**Pull request by asudarsa**
This commit reintroduces the following LLVM open-source PR: https://reviews.llvm.org/D145770
Some changes on top of the community change have been made as well.

Thanks

### CODE REVIEW

**Comment by asudarsa** [comment]
Keeping this in draft mode till testing passes. Thanks

**Comment by asudarsa** [comment]
Two linux failures are unrelated to this PR. Investigating the windows failure. Previously the test was marked with
// REQUIRES: powerpc-registered-target which would have prevented this test from running on our x86 targets. So, this failure might have been present without the proposed change as well.

Will investigate further.

Thanks

**Comment by asudarsa** [comment]
SYCL :: KernelAndProgram/kernel-bundle-merge-options-env.cpp is passing now.

File: clang/test/Driver/clang-offload-bundler-old-bundle-with-new-bundler.c [hunk 1]
```diff
@@ -0,0 +1,10 @@
+// REQUIRES: x86-registered-target
+// UNSUPPORTED: target={{.*}}-darwin{{.*}}, target={{.*}}-aix{{.*}}, system-windows
```
Comment (mdtoguchi) [inline clang/test/Driver/clang-offload-bundler-old-bundle-with-new-bundler.c:2]: Why won't this test work for Windows?

File: clang/lib/Driver/Driver.cpp [hunk 2]
```diff
@@ -5981,7 +5991,7 @@ class OffloadingActionBuilder final {
           SectionTriple += "-";
           SectionTriple += SyclTarget.BoundArch;
         }
-
+        SectionTriple = standardizedTriple(SectionTriple);
```
Comment (mdtoguchi) [inline clang/lib/Driver/Driver.cpp:5994]: If the bundler is doing the 'standardizing' to determine if the bundled sections are there - why do we need to do any standardizing on the caller side?

Comment (asudarsa) [reply clang/lib/Driver/Driver.cpp:5994]: Here, we are not calling the bundler. We seem to be reading the sections created by bundler directly.

Thanks

**Review by elizabethandrews — COMMENTED** [review]
Was this change reverted in community? Or did you just revert it in syclos to add modifications?

**Comment by asudarsa** [comment]
> Was this change reverted in community? Or did you just revert it in syclos to add modifications?

This was reverted in syclos as several tests were failing  and we needed time to react to upstream changes.

Thanks

**Comment by asudarsa** [comment]
@intel/dpcpp-tools-reviewers 

Friendly ping.

Thanks

File: clang/lib/Driver/Driver.cpp (hunk 2 above)
Comment (mdtoguchi) [reply clang/lib/Driver/Driver.cpp:5994]: Ah, I see - so here we are scanning all of the objects/archives and retrieving the section names.  These names could be names that are created with the updated standardized bundler behavior, requiring us to fix up any names for comparisons on the Driver end.

**Review by mdtoguchi — APPROVED** [review]

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp [hunk 3]
```diff
@@ -74,6 +74,24 @@ static void reportError(Error E) {
   logAllUnhandledErrors(std::move(E), WithColor::error(errs(), ToolPath));
 }
 
+// clang-offload-bundler is currently generating a 'standardized' target triple.
+// This is of the following form - Kind-Triple-TargetID.
+// Triple's format - Architecture-Vendor-OS-Environment
+// This routine transforms the target triple specified by user as input to
+// clang-offload-deps to this 'standardized' format.
+static std::string standardizedTarget(std::string OrigTarget) {
```
Comment (maksimsab) [inline clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:82]: I wonder why is it different from the version from `Driver.cpp`?

**Review by maksimsab — CHANGES_REQUESTED** [review]

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp (hunk 3 above)
Comment (asudarsa) [reply clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:82]: Driver.cpp standardizes only the triple (middle part of the Target). Target here consists of three parts: 1. kind (sycl, host, openmp). 2. target triple (arch-vendor-os-env; not sure why they call it triple when they have 4 parts) and 3. Target ID (sycldevice, nvidia, etc). 

Thanks

File: clang/lib/Driver/Driver.cpp [hunk 4]
```diff
@@ -115,6 +115,16 @@ using namespace clang::driver;
 using namespace clang;
 using namespace llvm::opt;
 
+static std::string standardizedTriple(std::string OrigTriple) {
```
[the remark this answers is not in the document]
Comment (asudarsa) [reply clang/lib/Driver/Driver.cpp:123]: Thanks. Added comment.

**Comment by asudarsa** [comment]
Latest change only added comments. Test failures do not seem to be related to my changes.

Thanks

File: clang/tools/clang-offload-deps/ClangOffloadDeps.cpp (hunk 3 above)
Comment (maksimsab) [reply clang/tools/clang-offload-deps/ClangOffloadDeps.cpp:82]: Thank you

File: clang/lib/Driver/Driver.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (maksimsab) [reply clang/lib/Driver/Driver.cpp:123]: Thank you

**Review by maksimsab — APPROVED** [review]
LGTM.

**Review by sarnex — APPROVED** [review]
Discussed ESIMD emu test failures with Arvind offline, they are definitely not related to this PR and seem to be a CI issue. These tests should be disabled. I didn't review the rest of the change.

**Comment by asudarsa** [comment]
Hi @AlexeySachkov 

I have looked at all the failures and none of them are related to my PR. Can you please take a look at this PR and merge if appropriate?

Thanks
Sincerely

**Comment by AlexeySachkov** [comment]
> Hi @AlexeySachkov
> 
> I have looked at all the failures and none of them are related to my PR. Can you please take a look at this PR and merge if appropriate?
> 
> Thanks Sincerely

I'm a bit hesitant to merge the PR with so many failed tests. Let's take a look at newer results first to make sure that those failures are indeed unrelated

**Comment by AlexeySachkov** [comment]
Fails:

```
********************
Failed Tests (1):
  SYCL :: ESIMD/accessor_local.cpp
```

Covered by https://github.com/intel/llvm/issues/10138

```
********************
Failed Tests (4):
  SYCL :: USM/memops2d/copy2d_device_to_host.cpp
  SYCL :: USM/memops2d/copy2d_host_to_device.cpp
  SYCL :: USM/memops2d/memcpy2d_device_to_host.cpp
  SYCL :: USM/memops2d/memcpy2d_host_to_device.cpp
```

Covered by https://github.com/intel/llvm/issues/10157

**Comment by steffenlarsen** [comment]
@asudarsa - clang/test/Driver/clang-offload-bundler-standardize.c is failing in post-commit after this change. Could you please address this ASAP?

**Comment by asudarsa** [comment]
Looking at this now. It looks mostly like a issue in options being handled by driver.
 Thanks.

**Pull request merged**